kv: try next replica on RangeNotFoundError

tbg opened the pull request:
Previously, if a Batch RPC came back with a RangeNotFoundError, we would
immediately stop trying to send to more replicas, evict the range
descriptor, and start a new attempt after a back-off.

This new attempt could end up using the same replica, so if the
RangeNotFoundError persisted for some amount of time, so would the
unsuccessful retries for requests to it as DistSender doesn't aggressively
shuffle the replicas.

It turns out that there are such situations, and the election-after-restart
roachtest spuriously hit one of them:

1. new replica receives a preemptive snapshot and the ConfChange
2. cluster restarts
3. now the new replica is in this state until the range wakes
   up, which may not happen for some time. 4. the first request to the range
   runs into the above problem

@nvanbenschoten: I think there is an issue to be filed about the tendency
of DistSender to get stuck in unfortunate configurations.

Fixes #30613.

Release note (bug fix): Avoid repeatedly trying a replica that was found to
be in the process of being added.

tbg committed:
roachtest: print output on failure in election test

This was hiding the output if the invocation itself failed, which is
when you wanted it most.

Release note: None

tbg committed:
kv: improve a trace event about descriptor eviction

Release note: None

tbg committed:
storage: report optional StoreID with RangeNotFoundError

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31013)
<!-- Reviewable:end -->

nvb reviewed (dismissed):
Nice find! These kinds of issues are very tricky to track down.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 10 of 10 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-:-LO4KuL12WHW4EjQM_4-:b-ea5km9)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 1117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4GypwEwy_3wMF6o7S:-LO4GypwEwy_3wMF6o7T:bih8y7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/b853fe5df149168ba421e01079fae3443d3759eb/pkg/kv/dist_sender.go#L1117)):*
> ```Go
> 			// descriptor. Invalidate the cache and try again with the new
> 			// metadata.
> 			log.Eventf(ctx, "evicting range descriptor on %T and backoff for re-lookup: %+v", tErr, desc)
> ```

Should we make this a `VEventf`?
___
*[pkg/kv/dist_sender.go, line 1360 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4JawC2x0tsaBi_C_J:-LO4JawC2x0tsaBi_C_K:b6iu0jg) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1360)):*
> ```Go
> 				// These errors are likely to be unique to the replica that reported
> 				// them, so no action is required before the next retry.
> 			case *roachpb.RangeNotFoundError:
> ```

The effect of this change is that cases where all replicas return `RangeNotFoundErrors` will now result in a `SendError` once the transport is exhausted. I think that's ok because we treat them the same in `sendPartialBatch`, but I want to make sure that's your intention.
___
*[pkg/kv/dist_sender.go, line 1362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4IQcD5TYNucpreyOd:-LO4IQcD5TYNucpreyOe:b5tp76r) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1362)):*
> ```Go
> 			case *roachpb.RangeNotFoundError:
> 				// The store we routed to doesn't have this replica. This can happen when
> 				// our descriptor is outright outdated, but it can also be caused a replica
> ```

"caused by"
___
*[pkg/kv/dist_sender.go, line 1371 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4JrMQ8OhTXW7k8V1W:-LO4JrMQ8OhTXW7k8V1X:b-hq5rp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1371)):*
> ```Go
> 				// up (which may never happen if the target range is dormant).
> 				if tErr.StoreID != 0 {
> 					cachedStoreID, found := ds.leaseHolderCache.Lookup(ctx, tErr.RangeID)
> ```

In your test failure, was the replica that the request continued to get routed to the leaseholder? I'm wondering if we should propagate the error immediately if we think it is the leaseholder but it returns a `RangeNotFoundError`.
___
*[pkg/kv/dist_sender.go, line 1374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4Iaqw625xZrGJK0Hs:-LO4Iaqw625xZrGJK0Ht:bek28f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1374)):*
> ```Go
> 					match := cachedStoreID == tErr.StoreID
> 					evicting := found && match
> 					log.Eventf(ctx, "evicting leaseholder s%d for r%d after RangeNotFound: %t (found=%t, match=%t)", tErr.StoreID, tErr.RangeID, evicting, found, match)
> ```

Should this be moved a line down?
___
*[pkg/kv/dist_sender.go, line 1391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4J8bp8rW46Sxuxfh0:-LO4J8bp8rW46Sxuxfh1:b-71mhht) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1391)):*
> ```Go
> 					if replicas.FindReplica(lh.StoreID) == -1 {
> 						// Replace NotLeaseHolderError with RangeNotFoundError.
> 						br.Error = roachpb.NewError(roachpb.NewRangeNotFoundError(rangeID, curReplica.StoreID))
> ```

It's surprising that a `NotLeaseHolderError` will be converted to a `RangeNotFoundError` and propagated immediately but a `RangeNotFoundError` won't be propagated (see comment above). Should we change this to a SendError and remove the handling of `RangeNotFoundErrors` entirely from `sendPartialBatch`?
___
*[pkg/storage/store.go, line 144 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4HvBB5041soVUFTs5:-LO4HvBB5041soVUFTs6:b-3fkv92) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf7f0aac388cb0fc99ec9b37b6ad071fa469564/pkg/storage/store.go#L144)):*
> ```Go
> 		Settings:                    st,
> 		AmbientCtx:                  log.AmbientContext{Tracer: st.Tracer},
> 		Clock:                       clock,
> ```

Is this a gofmt 1.10 vs 1.11 thing?


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 1 of 1 files at r2, 7 of 10 files at r4, 1 of 1 files at r5, 1 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/election.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOFQSKHA3iCBAZLBRZd:-LOFQSKHA3iCBAZLBRZe:ber87av) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/cmd/roachtest/election.go#L60)):*
> ```Go
> 			// eagerly that multiple elections conflict with each other).
> 			start = timeutil.Now()
> 			buf, err := c.RunWithBuffer(ctx, c.l, c.Node(1), `env COCKROACH_CONNECT_TIMEOUT=120 ./cockroach sql --insecure -e "
> ```

Why so long? The point of this test is that the cluster should be able to recover quickly. 
___
*[pkg/kv/dist_sender.go, line 1391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4J8bp8rW46Sxuxfh0:-LOFkUKT-ybhZs20we58:b-4aqrsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1391)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

It's surprising that a `NotLeaseHolderError` will be converted to a `RangeNotFoundError` and propagated immediately but a `RangeNotFoundError` won't be propagated (see comment above). Should we change this to a SendError and remove the handling of `RangeNotFoundErrors` entirely from `sendPartialBatch`?
</blockquote></details>

+1 (I think, haven't thought through this part fully)


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-:-LOIRf2Y-EmpXk9-5MZu:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/election.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOFQSKHA3iCBAZLBRZd:-LOI7BfV9EC7h7Cp_nzT:bvhab91) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/cmd/roachtest/election.go#L60)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why so long? The point of this test is that the cluster should be able to recover quickly. 
</blockquote></details>

All I know is that 5s is too short. I assume this is because Gossip needs a second to bootstrap (we start the processes asynchronously). I'll take this out and we can look at that separately.
___
*[pkg/kv/dist_sender.go, line 1117 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4GypwEwy_3wMF6o7S:-LOI9WvD9HEoK339xIqP:b-1v4gz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b853fe5df149168ba421e01079fae3443d3759eb/pkg/kv/dist_sender.go#L1117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we make this a `VEventf`?
</blockquote></details>

Sure, why not. Doubt it matters, but doesn't hurt.
___
*[pkg/kv/dist_sender.go, line 1360 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4JawC2x0tsaBi_C_J:-LOI7YdB9CzszLf7MIA0:b-gxhc2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1360)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The effect of this change is that cases where all replicas return `RangeNotFoundErrors` will now result in a `SendError` once the transport is exhausted. I think that's ok because we treat them the same in `sendPartialBatch`, but I want to make sure that's your intention.
</blockquote></details>

Yes, I think that's fine.
___
*[pkg/kv/dist_sender.go, line 1362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4IQcD5TYNucpreyOd:-LOI7kJQB8BSze-aMKNI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1362)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"caused by"
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 1371 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4JrMQ8OhTXW7k8V1W:-LOI7no8CmrqnWPf7cgQ:bfgp7jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1371)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In your test failure, was the replica that the request continued to get routed to the leaseholder? I'm wondering if we should propagate the error immediately if we think it is the leaseholder but it returns a `RangeNotFoundError`.
</blockquote></details>

No, in my testing this code never fired. But if it were cached as leaseholder we'd try it first on every subsequent attempt. If the real leaseholder were, say, the second replica, we'd never evict the cache and always pay a round-trip to that first node.

The real fix here is updating the leaseholder cache on successful replies, but I want to keep this diff small.
___
*[pkg/kv/dist_sender.go, line 1374 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4Iaqw625xZrGJK0Hs:-LOI8EHRDkbtMD_SO_pt:b1cic44) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1374)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this be moved a line down?
</blockquote></details>

This message was in the right place when I needed it (since I wanted to know what was going on) but now it seems right to move it. Done.
___
*[pkg/kv/dist_sender.go, line 1391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4J8bp8rW46Sxuxfh0:-LOI99fm3eMzXqhcZum1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1391)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

+1 (I think, haven't thought through this part fully)
</blockquote></details>

Done.
___
*[pkg/storage/store.go, line 144 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4HvBB5041soVUFTs5:-LOI9AgH7MpbxW3Mg5eR:b-q3pnco) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf7f0aac388cb0fc99ec9b37b6ad071fa469564/pkg/storage/store.go#L144)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this a gofmt 1.10 vs 1.11 thing?
</blockquote></details>

Not sure, but CI doesn't like it, so removed.


<!-- Sent from Reviewable.io -->

tbg committed:
kv: try next replica on RangeNotFoundError

Previously, if a Batch RPC came back with a RangeNotFoundError,
we would immediately stop trying to send to more replicas, evict the
range descriptor, and start a new attempt after a back-off.

This new attempt could end up using the same replica, so if the
RangeNotFoundError persisted for some amount of time, so would the
unsuccessful retries for requests to it as DistSender doesn't
aggressively shuffle the replicas.

It turns out that there are such situations, and the
election-after-restart roachtest spuriously hit one of them:

1. new replica receives a preemptive snapshot and the ConfChange
2. cluster restarts
3. now the new replica is in this state until the range wakes
   up, which may not happen for some time.
4. the first request to the range runs into the above problem

@nvanbenschoten: I think there is an issue to be filed about the
tendency of DistSender to get stuck in unfortunate configurations.

Fixes #30613.

Release note (bug fix): Avoid repeatedly trying a replica that was found
to be in the process of being added.

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/election.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOFQSKHA3iCBAZLBRZd:-LOJ2d4156AndV2TFCAs:b-40p1b7) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/cmd/roachtest/election.go#L60)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

All I know is that 5s is too short. I assume this is because Gossip needs a second to bootstrap (we start the processes asynchronously). I'll take this out and we can look at that separately.
</blockquote></details>

OK. We should probably increase the default connect timeout then. 5s is too short, since connections in the real world may need to wait for the lease timeout for the cluster to become live. (Also I see that that env var is defined as a String interpreted as integer seconds when it should probably be a Duration instead)
___
*[pkg/storage/store.go, line 144 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4HvBB5041soVUFTs5:-LOJ2tXsBXejbqZKeL7U:b1ashh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf7f0aac388cb0fc99ec9b37b6ad071fa469564/pkg/storage/store.go#L144)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Not sure, but CI doesn't like it, so removed.
</blockquote></details>

Yes, it is gofmt 1.11. You can configure your editor to use cockroach/bin/gofmt or crlfmt to make sure it uses the right version. 


<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ack. Holding off on the merge since https://github.com/cockroachdb/cockroach/issues/31068 fails like clockwork on this PR in CI.

tbg committed:
kv: cache leaseholder on successful response

Whenever a successful response is received from an RPC that we know has
to contact the leaseholder to succeed, update the leaseholder cache.

The immediate motivation for this is to be able to land the preceding
commits, which greatly exacerbated (as in, added a much faster failure
mode to)

```
make stress PKG=./pkg/sql/logictest TESTS=TestPlannerLogic/5node-dist/distsql_interleaved_join
```

However, the change is one we've wanted to make for a while; our caching
and in particular the eviction of leaseholders has been deficient
essentially ever since it was first introduced.

Touches #31068.

Release note: None

tschottdorf commented:
This should be green soon. I added a commit that populates the leaseholder cache on successful responses, PTAL!
I'm moderately excited about that commit because we've been wanting it for a while and it was actually easy. I'm less excited about the point in the cycle at which it's happening.

nvb reviewed (changes requested):
> Reviewed 11 of 11 files at r7, 1 of 1 files at r8, 10 of 10 files at r9, 3 of 3 files at r10, 5 of 5 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 1391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4J8bp8rW46Sxuxfh0:-LOOvBHI4_Y8_qQxUACM:b-lh45ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1391)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The other half of this is that we no longer need a case for `RangeNotFoundError` in `sendPartialBatch` at all. Is there a blocker that's preventing us from removing that now?
___
*[pkg/kv/dist_sender.go, line 1371 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOvlGa0tBAV9sLoTi0:-LOOvlGa0tBAV9sLoTi1:b-9kdj5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1371)):*
> ```Go
> 				// When a request that we know could only succeed on the leaseholder comes
> 				// back as successful, make sure the leaseholder cache reflects this
> 				// replica. In steady state, this should almost never be the case, and so we
> ```

This wording is a bit strange. This should _almost always_ be the case, which is why we gate it, right?
___
*[pkg/kv/dist_sender.go, line 1386 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOvtf85-bTgmRFYwuh:-LOOvtf85-bTgmRFYwui:b-lo8t96) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1386)):*
> ```Go
> 				// replica that has just been added but needs a snapshot to be caught up.
> 				//
> 				// We'll try other replicas which typically gives us the leaseholder, either
> ```

Glad we could remove that code.
___
*[pkg/kv/dist_sender.go, line 1396 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOw2Y03jOCHT4SeMmM:-LOOw2Y03jOCHT4SeMmN:b-gu3bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1396)):*
> ```Go
> 					// the range descriptor), but that's OK - this kind of error should be rare
> 					// but when it happens, we want to populate the cache with something useful
> 					// ASAP without waiting for another round trip.
> ```

Can we change the `leaseHolder` object to avoid this?
___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOOwIOD0hW7rFcmtDPB:b918bht) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
> ```
> /5         NULL     6         {5}       5
> 
> # Verify data placement.
> ```

Expand on this comment a bit. Why does the output change?


<!-- Sent from Reviewable.io -->

a-robinson reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOOzRc04qHPOnwvfDU-:begjb39) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Expand on this comment a bit. Why does the output change?
</blockquote></details>

It's a different table (`kv` vs `kw`).

And to answer @tschottdorf's original question in https://github.com/cockroachdb/cockroach/issues/31068#issuecomment-427887250, I imagine this makes the tests pass because it ensures that all the ranges/leases are in the DistSender's caches.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 1 of 11 files at r7, 8 of 10 files at r9, 1 of 3 files at r10, 5 of 5 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/kv/dist_sender.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOP6zzJC0d5TlC9GmQv:-LOP6zzJC0d5TlC9GmQw:b-w6shhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L433)):*
> ```Go
> }
> 
> type leaseHolderInfo struct {
> ```

These two things don't seem to fit together very well in one struct. I think it would make more sense to make them two separate parameters to sendToReplicas (or just recompute needsLeaseHolder on the fly from the batch when needed). 


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Ok, ready for a final look. Happy with how this turned out.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-:-LOUqFIy9SJvtmCGqmp4:b-rm93r)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/election.go, line 60 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOFQSKHA3iCBAZLBRZd:-LOThBEX8VrY4pGDQoVA:b-3iiuf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/cmd/roachtest/election.go#L60)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

OK. We should probably increase the default connect timeout then. 5s is too short, since connections in the real world may need to wait for the lease timeout for the cluster to become live. (Also I see that that env var is defined as a String interpreted as integer seconds when it should probably be a Duration instead)
</blockquote></details>

I'll keep this on the radar, but I'm going to leave as-is on the off-change that these timeouts were actually the same bug, hitting very early in one of the internal queries ran by `./cockroach sql`.
___
*[pkg/kv/dist_sender.go, line 1391 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LO4J8bp8rW46Sxuxfh0:-LOThnrt1ZYEhK1lHYBx:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/668da72496954c3bf90f4a843e7c08d7ee06c933/pkg/kv/dist_sender.go#L1391)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The other half of this is that we no longer need a case for `RangeNotFoundError` in `sendPartialBatch` at all. Is there a blocker that's preventing us from removing that now?
</blockquote></details>

Good point, done.
___
*[pkg/kv/dist_sender.go, line 433 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOP6zzJC0d5TlC9GmQv:-LOTlAe04VjjIhvcv5_1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L433)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These two things don't seem to fit together very well in one struct. I think it would make more sense to make them two separate parameters to sendToReplicas (or just recompute needsLeaseHolder on the fly from the batch when needed). 
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 1371 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOvlGa0tBAV9sLoTi0:-LOThVoT6An7pY8A0JZQ:b-p4ybu) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1371)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This wording is a bit strange. This should _almost always_ be the case, which is why we gate it, right?
</blockquote></details>

Oops, fixed.
___
*[pkg/kv/dist_sender.go, line 1386 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOvtf85-bTgmRFYwuh:-LOTiNJrA9HQwloU2PWR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1386)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Glad we could remove that code.
</blockquote></details>

Done.
___
*[pkg/kv/dist_sender.go, line 1396 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOw2Y03jOCHT4SeMmM:-LOTl9g492GLKRCrBL8Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/kv/dist_sender.go#L1396)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we change the `leaseHolder` object to avoid this?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOTi-nH2u6U-RwrF8WP:bnc6i4) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It's a different table (`kv` vs `kw`).

And to answer @tschottdorf's original question in https://github.com/cockroachdb/cockroach/issues/31068#issuecomment-427887250, I imagine this makes the tests pass because it ensures that all the ranges/leases are in the DistSender's caches.
</blockquote></details>

Yeah, my only explanation here is that without this query, some ranges may just never have been touched and so the leaseholders aren't cached.

On the flip side, I don't know why this lease placement here is deterministic. Do these tests apply constraints or something like that? I mentioned this trick to @andreimatei and his eyebrow raised visibly, so perhaps he wants to comment.

(I'm not planning to block this PR on that, though).


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 10 files at r14, 1 of 3 files at r15, 6 of 6 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-:-LOUwaia-d3juaR-1NAz:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!

bors r=nvanbenschoten,bdarnell

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOUyK7x81cHX8soDS_t:bgbc2jv) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, my only explanation here is that without this query, some ranges may just never have been touched and so the leaseholders aren't cached.

On the flip side, I don't know why this lease placement here is deterministic. Do these tests apply constraints or something like that? I mentioned this trick to @andreimatei and his eyebrow raised visibly, so perhaps he wants to comment.

(I'm not planning to block this PR on that, though).
</blockquote></details>

why wouldn't the placement be deterministic? Doesn't the `EXPERIMENTAL_RELOCATE` do deterministic moves? And after that runs, I assume there's  nothing to cause a change.


<!-- Sent from Reviewable.io -->

a-robinson reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOUzd-M5tCUujaPXaSO:bbergeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why wouldn't the placement be deterministic? Doesn't the `EXPERIMENTAL_RELOCATE` do deterministic moves? And after that runs, I assume there's  nothing to cause a change.
</blockquote></details>

Lease placement is deterministic because that's what `EXPERIMENTAL_RELOCATE` guarantees. It always transfers the lease to the first listed store, as specified in the original proposal https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20170314_sql_split_syntax.md#3-alter-tableindex-experimental_relocate


<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks for the clarifications, @a-robinson and @andreimatei.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOV0UwH98EmkplliJzn:b-iaablr) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Lease placement is deterministic because that's what `EXPERIMENTAL_RELOCATE` guarantees. It always transfers the lease to the first listed store, as specified in the original proposal https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20170314_sql_split_syntax.md#3-alter-tableindex-experimental_relocate
</blockquote></details>

right, so there's no mystery here, is there?
these `show experimental_ranges` queries could be replaced with `select count(1)` for the purposes of populating the caches, but I don't know if that'd be better or worse for the test.
I'm mildly surprised that `show experimental_ranges` populates any caches, but I don't know how it's implemented.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (approved):
> Reviewed 8 of 10 files at r14, 1 of 3 files at r15, 6 of 6 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

a-robinson reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31013)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/explain_analyze_plans, line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31013#-LOOwIOD0hW7rFcmtDPA:-LOV23RQBT883jRZzo08:bg5cl29) ([raw file](https://github.com/cockroachdb/cockroach/blob/faea6f47efd752cc3a92bfeeb54132d583105cf1/pkg/sql/logictest/testdata/logic_test/explain_analyze_plans#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

right, so there's no mystery here, is there?
these `show experimental_ranges` queries could be replaced with `select count(1)` for the purposes of populating the caches, but I don't know if that'd be better or worse for the test.
I'm mildly surprised that `show experimental_ranges` populates any caches, but I don't know how it's implemented.
</blockquote></details>

`show ranges` works for populating the caches because in order to populate the `lease_holder` column it has to send a `LeaseInfo` batch request to every range.


<!-- Sent from Reviewable.io -->
